pydicom/charset.py

```diff
@@ -95,6 +96,181 @@
+        encoded += b
+
+    return encoded
+
+
+def _encode_to_jis_x_0208(value, errors='strict'):
+    """Convert a unicode string into JIS X 0208 byte string using iso2022_jp
+    encodings.
+    The escape sequence which is located at the end of the encoded value has
+    to vary depending on the value 1 of SpecificCharacterSet. So we have to
+    trim it and append the correct escape sequence manually.
+
+    Parameters
+    ----------
+    value : text type
+        The unicode string as presented to the user.
+    errors : str
+        The behavior of a character which could not be encoded. This value
+        is passed to errors argument of str.encode().
+
+    Returns
+    -------
+    byte string
+        The encoded string. If some characters in value could not be encoded to
+        JIS X 0208, it depends on the behavior of iso2022_jp encoder.
+
+    Raises
+    ------
+    UnicodeEncodeError
+        If errors is set to 'strict' and `value` could not be encoded with
+        JIS X 0208.
+    """
+
+    # If errors is not strict, this function is used as fallback.
+    # So keep the tail escape sequence of encoded for backward compatibility.
+    if errors != 'strict':
+        return value.encode('iso2022_jp', errors=errors)
+
+    Encoder = codecs.getincrementalencoder('iso2022-jp')
+    encoder = Encoder()
+
+    encoded = encoder.encode(value[0])
+    if encoded[:3] != ENCODINGS_TO_CODES['iso2022_jp']:
+        raise UnicodeEncodeError(
+            'iso2022_jp', value, 0, len(value),
+            'Given character is out of ISO IR 87')
+
+    for i, c in enumerate(value[1:], 1):
+        try:
+            b = encoder.encode(c)
+        except UnicodeEncodeError as e:
+            e.start = i
+            e.end = len(value)
+            raise e
+        if b[:3] == ENCODINGS_TO_CODES['iso8859']:
+            raise UnicodeEncodeError(
+                'iso2022_jp', value, i, len(value),
+                'Given character is out of ISO IR 87')
+        encoded += b
+    return encoded
+
+
+def _get_escape_sequence_for_encoding(encoding, encoded=None):
+    """ Return an escape sequence corresponding to the given encoding. If
+    encoding is 'shift_jis', return 'ESC)I' or 'ESC(J' depending on the first
+    byte of encoded.
+
+    Parameters
+    ----------
+    encoding : str
+        An encoding is used to specify  an escape sequence.
+
+    encoded : bytes or str
+        The encoded value is used to chose an escape sequence if encoding is
+        'shift_jis'
+
+    Returns
+    -------
+    string
+        Escape sequence for encoded value.
+    """
+
+    ESC_ISO_IR_14 = ESC + b'(J'
+    ESC_ISO_IR_13 = ESC + b')I'
+
+    if encoding == 'shift_jis':
+        if encoded is None:
+            return ESC_ISO_IR_14
+
+        first_byte = encoded[0]
+        if isinstance(encoded, str):
+            first_byte = ord(first_byte)
+        if 0x80 <= first_byte:
```

Comment:
There use of  `str` without python 2/3 differentiation makes me nervous.  Is there a way to separate using explicit `if` statement on the `config.in_py2` variable for later removal when we drop python 2?

Reply:
Well, the statement is needed in Python 3, because `encoded` can by `bytes` or `str`. But it may be cleaner to use `isinstance(encoded, compat.string_types)`, so it would also cover `encoded` being `unicode`, though I'm not sure if this is possible here.

Reply:
Ok, this is correct - it is always `bytes` in Python 3, I wasn't sure about this. Thanks!

Reply:
@darcymason 
I modified to use config.in_py2. To take dropping python2 support into account, I think this is cleaner.

@mrbean-bremen
Could you review them again? The aim of these codes is to get the code value of the first character. I think that the str.encode() in python2 returns str value and str.encode() in python3 returns bytes value. So the 'if' section was expected to run in the python2 environment as far as I know.

Reply:
Yes, this is what I meant in my last comment - this looks good to me now!

Reply:
@mrbean-bremen
Thanks for your review! I was surprised at your quick response :)